Bug fixing for shape processing and inserting linked images

msamsel opened the pull request:
## What is the purpose of this pull request?

Bug fix

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What changes did you make?
- fix for 2 significant problem. Introduce both changes, because both contains modification of PFW filter and partially common fragment (`v:shape` element).
  - problem with processing grouped and adjacent shapes
    1. There is change in PFW filter to tag img which are shapes with additional `data` attribute.
    2. Next in PFW Image was add recognition of such attribute. This allows on ignore img tags belong to shapes.
    3. Also it was possible to simplify RTF parser to detect only images in it.
    4. And a little clean up was required to fix test, because after parser change shapes are not returned from RTF.
  - pasting images with link twice:
    1. There is change in PFW filter for `v:shape` processing.
    2. Old behaviour (which is probably left for IE8 compatibility, but I have not 100% sure about it) was triggered always
    3. Now it's triggered for shapes without `o:gfxdata` attribute. Only shapes poses this attribute, so in case of processing image old mechanism will be preserved. Alternative might be browser detection.
    4. In old behaviour was changed way how sibling tags are process. Now are searched recursively. In case of linked image, we don't have directly an `img` tag, but `span` tag which contains adequate `img` tag.
    5. This prevent of inserting images in Chrome and Firefox.

close #995 , close #1069

msamsel commented on tests/plugins/pastefromword/generated/_fixtures/Shape_adjacent_image/expected.html:
```diff
@@ -0,0 +1 @@
+<p style="margin-left:0in; margin-right:0in"></p><table style="width:100%"><tbody><tr><td><div><p>hello world</p></div></td></tr></tbody></table><p><br /></p><p><br /></p><table align="left"><tbody><tr><td><br /></td><td><br /></td><td><br /></td></tr><tr><td><br /></td><td colspan="2" valign="top"><img data-cke-is-shape="true" src="file:///c:\users\foobar\appdata\local\temp\msohtmlclip1\01\clip_ ...
```
This test work a little bit different.
1. Textbox is treat as a shape. So there is 4 shapes in source doc.
2. Because 3 first shapes are adjacent, they are treat as a single img in the output. Take a look here: https://github.com/ckeditor/ckeditor-dev/blob/534af88e2b35d7e4906c38b44957940d77e2c5d8/tests/plugins/pastefromword/generated/_fixtures/Shape_adjacent_image/word2013/chrome.html#L1084
It shows that this single `img` tag belong to 3 shapes with given names.
3. If you remove textbox and remain other shapes untouched, this will generate 3 img tags in the output, because 3 top shapes will stop being adjacent.

f1ames replied:
Makes sense, haven't noticed that detail. So this means the expected output is valid in this case, isn't it? 👍

msamsel replied:
I think it's valid here :)

msamsel commented:
* I've update test cases with Chrome browser from MacOS. Firefox has exactly the same markup so there is no need to duplicate it and thats why is missing in tests.
* `Canvas` test case doesn't return `img` tag in `text/html` clipboard on MacOS. That's why those aren't present in the test cases. Having test without img tag in output, where Windows have 4 of them, might be confusing for analysing this test case it in the future.
* There is update in condition which apply old behaviour of `v:shape` parsing. Unfortunately simple checking of `CKEDITOR.env` doesn't allow on proper work of autogenerated tests. Right now `v:shape` (without `o:gfxdata` attribute) can be processed with an old way (IE8 way), if fulfil 2 additional conditions. Those condition are required to detect `canvas` usage in `text/html` markup. Canvas generate empty `v:shapes` which should not be transform into `img` with an old mechanism.
  - https://github.com/ckeditor/ckeditor-dev/blob/aee832046414f46a3ddea923c3c696bfe5513293/plugins/pastefromword/filter/default.js#L397
  - https://github.com/ckeditor/ckeditor-dev/blob/aee832046414f46a3ddea923c3c696bfe5513293/plugins/pastefromword/filter/default.js#L402-L406

msamsel commented:
I fix those minor stuff and rename test. When I start to analyse canvas issue and try to describe it as separate issue I found possible solution. It seem like changing one `if` solve the case. Currently canvas shapes are properly tagged. :)

f1ames commented on plugins/pastefromword/filter/default.js:
```diff
@@ -363,31 +392,49 @@
+
+						// Canvas leaves empty `v:shape`, which should not be converted into img tag.
+						// These empty `v:shape` contains 2 attributes which helps distinguish it.
+						child = element.getFirst( 'v:imagedata' );
```
This can be assigned on initialization, 5 lines above. And then the comment will fit better, right after `if` which it describes.

msamsel replied:
I prefer not to move comment after `if, as it might imply relation to `shapeTagging` method. All other `if` description in entire file seems to be described just before it.

f1ames replied:
@msamsel I fully agree, my mistake here, what I meant was 

> And then the comment will fit better, right **before** if which it describes

By merging `var` initialization you achieved just what I was thinking about so it is ok 👍

msamsel commented on tests/plugins/pastefromword/manual/shapesaretagged.md:
```diff
@@ -0,0 +1,11 @@
+
+**Expected:**
+  * For browsers with good `text/html` clipboard processing (Firefox, Chrome, Safari): First image tag had additional attribute `data-cke-is-shape="true"`. Second image is pasted without this attribute.
+  * For other browsers (IE, Mobile): There won't be `data-cke-is-shape="true"`. Browser will behave in an old way: display blank place or embed image by it's own.
```
I forgot that Edge exists :D.

msamsel commented on tests/plugins/pastefromwordimage/manual/pasteimagewithlink.md:
```diff
@@ -0,0 +1,16 @@
+### Perform above steps for all files:
+  1. [Link_image.docx](../generated/_fixtures/Link_image/Link_image.docx)
+
+**Expected:** Images is visible and pasted only once. Image is wrapped with an anchor link.
```
It appear that `image` plugin was missing in those test. Because PFW Image register img tag as allowed, the img appear but had stripped away width and height.

msamsel commented:
I've added an issue on our side to remember about checking Safari in future: #1134.
I've replaced ignore condition:
from:
```
!CKEDITOR.plugins.clipboard.isCustomDataTypesSupported || CKEDITOR.env.iOS
```
to:
```
!CKEDITOR.plugins.clipboard.isCustomDataTypesSupported || CKEDITOR.env.safari
```
iOS is tested inside `isCustomDataTypesSupported`, so there is no need to double check it here.
